### PR TITLE
feat: Added AnchorToEffect and AnchorByEffect

### PR DESCRIPTION
**Pull request by st-pasha**
# Description

- Added `AnchorProvider` interface, currently implemented by `PositionComponent` and `Viewfinder`.
- Added `AnchorToEffect` and `AnchorByEffect` that can be applied to a target which is an `AnchorProvider`.

## Checklist

- [x] The title of my PR starts with a [Conventional Commit] prefix (`fix:`, `feat:`, `docs:` etc).
- [x] I have read the [Contributor Guide] and followed the process outlined for submitting PRs.
- [x] I have updated/added tests for ALL new/updated/fixed functionality.
- [x] I have updated/added relevant documentation in `docs` and added dartdoc comments with `///`.
- [ ] I have updated/added relevant examples in `examples`.

## Breaking Change

- [-] Yes, this is a breaking change.
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flame-engine/flame/issues
[Contributor Guide]: https://github.com/flame-engine/flame/blob/main/CONTRIBUTING.md
[Flame Style Guide]: https://github.com/flame-engine/flame/blob/main/STYLEGUIDE.md
[Conventional Commit]: https://conventionalcommits.org

### CODE REVIEW

**Review by spydon — APPROVED** [review]

**Review by wolfenrain — APPROVED** [review]
LGTM

File: packages/flame/lib/src/anchor.dart [hunk 1]
```diff
@@ -113,7 +113,7 @@ class Anchor {
 
   @override
   bool operator ==(Object other) {
-    return other is Anchor && hashCode == other.hashCode;
+    return other is Anchor && x == other.x && y == other.y;
```
[the remark this answers is not in the document]
Comment (st-pasha) [reply packages/flame/lib/src/anchor.dart:116]: `hashCode` is an int, which is 32 bit in Dart; `x` and `y` are both doubles, so together they have 128 bits. This means that on average there are 2^96 different pairs of `(x,y)` that map to the same hash code; which is to say for every anchor there are 2^96 other (different) anchors that would compare equal.

**Comment by st-pasha** [comment]
It just occurred to me that `AnchorByEffect` will also be useful to have.

**Comment by erickzanardo** [comment]
Can you share a use case for this? Just really curious, never thought myself on animating a component anchor

**Comment by st-pasha** [comment]
We already have Anchor effects in the current camera, only it's called `relativeOffset` there, and when you call `setRelativeOffset(Anchor newRelativeOffset)`, it applies a smooth transition of the anchor. So, `AnchorToEffect()` is an equivalent mechanism for the new camera, but allows a finer control over the timing of the transition.

Then, there is camera shake. It can also be done via an effect -- either affecting the camera's position, or the camera's anchor. It occurs to me that doing the shake via an anchor effect would be more reliable, since the camera's position is already affected by various behaviors such as follow-component of world-boundary. So, this is where `AnchorEffect.by` would come in: add a tiny amount of random offset to the camera's anchor, and it shakes nicely without any interference from the position behaviors.

File: packages/flame/lib/src/anchor.dart (hunk 1 above)
[the remark this answers is not in the document]
Comment (spydon) [reply packages/flame/lib/src/anchor.dart:116]: Oh right! Forgot that it was an int.
Well to your question @wolfenrain, we can either add it manually to the changelog when we do the release or it should be done as two PRs.

[the remark this answers is not in the document]
Comment (st-pasha) [reply packages/flame/lib/src/anchor.dart:116]: I moved this to a separate PR #1560.

**Pull request merged**